File: kotlin/src/main/kotlin/377.CombinationSumIV.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/combination-sum-iv/


class CombinationSumIV{
  fun combinationSum4(nums: IntArray, target: Int): Int {
    val dp = IntArray(target+1)
    dp[0] = 1
    for(t in 1..target)
      for(v in nums)
        if(t-v >= 0)
          dp[t] += dp[t-v]
    return dp[target]
  }

  @Test
  fun test1(){
    assertEquals(7, combinationSum4(intArrayOf(1,2,3), 4))
  }
}

File: python/377_CombinationSumIV.py
from unittest import TestCase
# https://leetcode.com/problems/combination-sum-iv/


class CombinationSumIV(TestCase):
    def combinationSum4(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        dp = [1, *[0]*target]
        for t in range(1, target+1):
            for num in nums:
                if t-num >= 0:
                    dp[t] += dp[t-num]
        return dp[target]

    def test1(self):
        self.assertEqual(7, self.combinationSum4([1,2,3], 4))
